snuba/datasets/schemas/__init__.py

```diff
@@ -1,9 +1,29 @@
+from __future__ import annotations
+
 from abc import ABC, abstractmethod
 from typing import Mapping, List
 
 from snuba.clickhouse.columns import ColumnSet
 
 
+class RelationalSource(ABC):
+    """
+    Represent an abstract view over a relational datasource, which
+    can represent a table or a join.
+    """
+
+    @abstractmethod
+    def copy(self) -> RelationalSource:
+        raise NotImplementedError
+
+    @abstractmethod
+    def format(self) -> str:
```

Comment:
I did not just define __str__ intentionally because this method will take a parameter in the future. I am thinking of implementing a Strategy Pattern for this: the abstract representation of the query knows how to build a formatted SQL query in abstract terms, then we will have a CLickhouseStrategy that provides the clickhouse specific features.

This allows us to prevent ClickhouseQuery to have full access into the state of Query, which would have seemed a leaky abstraction.